Add a version command to neo4j-admin

spacecowboy opened the pull request:
Help text is as follows:

```
usage: neo4j-admin version --store=<path-to-dir>

Checks the version of a Neo4j database store. Note that this command expects a
path to a store directory, for example --store=data/databases/graph.db.

options:
  --store=<path-to-dir>   Path to database store to check version of.
```

changelog [tools] `neo4j-admin`: added `version` command

This command can be pointed against a store and reports information
about its version. 

```
usage: neo4j-admin version --store=<path-to-dir>
```

Two example outputs are

- If store matches latest known version to the tool:

```
Store format version:    v0.A.7
Introduced in version:   3.0.0
Current version:         3.0.5
```

- If store is older than latest known version to the tool:

```
Store format version:    v0.A.5
Introduced in version:   2.2.0
Superseded in version:   2.3.0
Current version:         3.0.5
```

spacecowboy committed:
Add method which tells which Neo4jversion format was introduced in

spacecowboy committed:
Create a method which can read store version of store

spacecowboy committed:
Add `version` command to `neo4j-admin`

This command can be pointed against a store and reports information
about its version. Two example outputs are

- If store matches latest known version to the tool:

```
Store version:      v0.A.7
Introduced in:      3.0.0
```

- If store is older than latest known version to the tool:

```
Store version:      v0.A.5
Introduced in:      2.2.0
Superceded in:      2.3.0
```

Help text is as follows:

```
usage: neo4j-admin version --store=<path-to-dir>

Checks the version of a Neo4j database store. Note that this command expects a
path to a store directory, for example --store=data/databases/graph.db.

options:
  --store=<path-to-dir>   Path to database store to check version of.
```

spacecowboy committed:
Add enterprise test for version command

spacecowboy committed:
Move findSuccessor method into RecordFormatSelector

spacecowboy commented:
@MishaDemianenko I have addressed your initial comments

MishaDemianenko commented on community/dbms/src/main/java/org/neo4j/commandline/dbms/VersionCommand.java:
```diff
@@ -0,0 +1,122 @@
+            Optional<RecordFormats> nextFormat = findSuccessor( format );
+            if ( nextFormat.isPresent() )
+            {
+                out.accept( String.format( fmt, "Superceded in:", nextFormat.get().neo4jVersion() ) );
```
typo: Superseded

MishaDemianenko commented on community/dbms/src/main/java/org/neo4j/commandline/dbms/VersionCommand.java:
```diff
@@ -0,0 +1,122 @@
+            out.accept( String.format( fmt, "Introduced in:", format.neo4jVersion() ) );
+
+            Optional<RecordFormats> nextFormat = findSuccessor( format );
+            if ( nextFormat.isPresent() )
```
What use map instead and say that format is superseded or the latest one (still newest)?

MishaDemianenko commented on community/dbms/src/main/java/org/neo4j/commandline/dbms/VersionCommand.java:
```diff
@@ -0,0 +1,122 @@
+            final String storeVersion = new StoreVersionCheck( pageCache )
+                    .getVersion( storeDir.resolve( MetaDataStore.DEFAULT_NAME ).toFile() )
+                    .orElseThrow(
+                            () -> new CommandFailed( String.format( "Could not find a version in '%s'", storeDir ) ) );
```
I find this message a bit cryptic, what about: Could not find version in database metadata file in store _store_ or something similar?

MishaDemianenko commented on community/dbms/src/test/java/org/neo4j/commandline/dbms/VersionCommandTest.java:
```diff
@@ -0,0 +1,192 @@
+    @Before
+    public void setUp() throws Exception
+    {
+        fs = new DefaultFileSystemAbstraction();
```
We also have junit rules for file system abstraction and page cache: DefaultFileSystemRule and PageCacheRule

MishaDemianenko commented on community/dbms/src/test/java/org/neo4j/commandline/dbms/VersionCommandTest.java:
```diff
@@ -0,0 +1,192 @@
+        databaseDirectory = homeDir.resolve( "data/databases/foo.db" );
+        Files.createDirectories( databaseDirectory );
+        baos = new ByteArrayOutputStream();
+        PrintStream printStream = new PrintStream( baos );
```
Couple of thing here: you creating print stream without auto flushing, use  ByteArrayOutputStream.toString which will use system dependent encoding to produce final comparable string. What do you think if we will create rule for us that will gonna provide assertable string consumer instead where we will fix all those comments?

MishaDemianenko commented on enterprise/neo4j-enterprise/src/test/java/org/neo4j/commandline/dbms/VersionCommandEnterpriseTest.java:
```diff
@@ -0,0 +1,111 @@
+        Path homeDir = testDirectory.directory( "home-dir" ).toPath();
+        databaseDirectory = homeDir.resolve( "data/databases/foo.db" );
+        Files.createDirectories( databaseDirectory );
+        baos = new ByteArrayOutputStream();
```
same comments as above

MishaDemianenko commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/highlimit/v300/HighLimitV3_0_0.java:
```diff
@@ -68,6 +68,12 @@ public HighLimitV3_0_0()
     @Override
+    public String neo4jVersion()
+    {
+        return "3.0.0";
```
what instead of store this in each format we will add a field in StoreVersion (introducedIn field) for example and then we will just return that field value from each of the format?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/RecordFormats.java:
```diff
@@ -49,6 +49,12 @@ public Factory( String key, String... altKeys )
+     * @return the neo4j version where this format was introduced. It is almost certainly NOT the only version of
+     * neo4j where this format is used.
+     */
+    String neo4jVersion();
```
should this be something more like introduced in version?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreVersionCheck.java:
```diff
@@ -38,32 +39,45 @@ public StoreVersionCheck( PageCache pageCache )
+        if ( storeVersion.isPresent() )
         {
-            return new Result( Outcome.storeVersionNotFound, null, neostoreFile.getName() );
+            if ( expectedVersion.equals( storeVersion.get() ) )
```
looks like nice place to use `map`/`filter` operation instead

MishaDemianenko reviewed (changes requested):
Couple of comments

spacecowboy committed:
Address review comments and make code more functional

Also moves the location of first neo4j version string so that formats do
not have to implement that method.

spacecowboy commented:
@MishaDemianenko please have another look at your earliest convenience. I think I addressed all the points.

There is one test left with a`ByteArrayOutputStream` which I want to keep, because I really want to test for the line-endings in that case. But I think that's fine since `flush()` is a `noop` on `ByteArrayOutputStream`.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/StoreVersion.java:
```diff
@@ -24,25 +24,32 @@
+
+    public String firstNeo4jVersion()
+    {
+        return firstNeo4jVersion;
```
i would really like if we will call this version `introductionVersion` or `introducedInVersion`. First version is not descriptive by itself and as soon as we use `first` it mean that we have some order in mind which is not clear.

MishaDemianenko commented on community/dbms/src/main/java/org/neo4j/commandline/dbms/VersionCommand.java:
```diff
@@ -0,0 +1,122 @@
+            out.accept( String.format( fmt, "Store version:", storeVersion ) );
+
+            RecordFormats format = RecordFormatSelector.selectForVersion( storeVersion );
+            out.accept( String.format( fmt, "Introduced in:", format.firstNeo4jVersion() ) );
```
i think if we will change messages to: `Introduced in version:`, `Superseded in version:` and `Current version:` that will be more clear. WDYT?

MishaDemianenko reviewed (commented):
2 comments, all the rest looks great!

spacecowboy committed:
Rename to introductionVersion and clarify output of version command

spacecowboy commented:
@MishaDemianenko I agree, changed according to feedback
